### PR TITLE
C domain, fix performance regression since 3.0.0

### PR DESCRIPTION
**Pull request by donaldh**
### Feature or Bugfix
- Bugfix

### Purpose

Sphinx 3.0.0 onwards have a significant performance regression from 2.4.4 for the C domain. The cProfile output shows that this is due to _find_named_symbols using a linear search. Here's the profile information for the Linux kernel 'make htmldocs':

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
770364892  165.102    0.000  165.102    0.000 sphinx/domains/c.py:153(__eq__)
   104124  163.968    0.002  544.788    0.005 sphinx/domains/c.py:1731(_find_named_symbols)
543888397  123.767    0.000  176.685    0.000 sphinx/domains/c.py:1679(children_recurse_anon)
     4292   74.081    0.017   74.081    0.017 {method 'poll' of 'select.poll' objects}
631233096   69.389    0.000  246.017    0.000 sphinx/domains/c.py:1746(candidates)
121406721/3359598   65.689    0.000   76.762    0.000 docutils/nodes.py:202(_fast_findall)
  3477076   64.387    0.000   65.758    0.000 sphinx/util/nodes.py:633(_copy_except__document)
544032973   52.950    0.000   52.950    0.000 sphinx/domains/c.py:156(is_anon)
79012597/3430   36.395    0.000   36.395    0.011 sphinx/domains/c.py:1656(clear_doc)
286882978   31.271    0.000   31.279    0.000 {built-in method builtins.isinstance}

Function                                was called by...
                                           ncalls  tottime  cumtime
sphinx/domains/c.py:153(__eq__)      <- 631153346  134.803  134.803  sphinx/domains/c.py:1731(_find_named_symbols)
                                           154878    0.041    0.041  sphinx/domains/c.py:2085(find_identifier)
                                        139056533   30.259   30.259  sphinx/domains/c.py:2116(direct_lookup)
                                              135    0.000    0.000  sphinx/util/cfamily.py:89(__eq__)
```
Replace `Symbol._children` with dicts, one keyed by `ident` and one keyed by `docname`. This lets us remove `_find_named_symbols` and replace these usage patterns with fast child lookup. Also use the `_anonChildren` list to speed up searching through anonymous children.

Note that dict is guaranteed to maintain insertion order since Python 3.7 so we retain iteration. Whenever iteration is required, we use _childrenByName.values().

Before – with Sphinx 7.2.6:
```
% time make htmldocs
...
real	9m0.533s
user	15m38.397s
sys	1m0.907s
```
After - with Sphinx 7.3.0+:
```
% time make htmldocs
...
real	4m27.063s
user	10m54.985s
sys	0m57.702s
```

### CODE REVIEW

**Comment by picnixz** [comment]
(I will have comments on that tomorrow, so I'll ask other maintainers not to merge this)

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Update this comment.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
    def __init__(self, symbols: Sequence[Symbol], parentSymbol: Symbol,
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: I feel this requires a method.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Should we check whether `self` is in the `childrenByDocname[self.docname]` as well? (thus the needs of a method that does this for you, e.g., ``self.parent._addChild(self.docname, self)``

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: I would suggest adding to the `ASTIdentifier` a property alias `self.ident.name` (currently it's called `identifier` but `self.ident.identifier` is clearly not good).

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
```

Before, `remove` would have raise a ValueError but wouldn't be suppressed with `-O`. So I think we need to keep (possible) exceptions if we do `del`.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Do you need a list of symbols or can you live with a set? (same question for `_anonChildren`).

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Shouldn't it be

```suggestion
        siblings = self.parent._childrenByDocname.get(self.docname, [])
        if siblings and self in siblings:
            siblings.remove(self)
        # next two lines are optional, I don't know if it's better or not
        if not siblings: 
            self.parent._childrenByDocname.pop(self.docname, None)
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        for sChild in self._children:
```

(After you have created `self._children = self._childrenByName.values()`)

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        for c in self._children:
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        return ''.join([self.to_string(indent), *(c.dump(indent + 1) for c in self.children)])
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
			res = s._childrenByName.get(str(ident))
```

and remove the `res = None` above (I cannot suggest on non-modified lines).

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Make it a method say `_find_anonymous_child`

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        for otherChild in other.children:
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [reply sphinx/domains/c/_symbol.py]: You can also add a private property ``_children`` which is an alias to `self._childrenByName.values()` (and since it's a proxy, it'll be updated whenever the mapping is updated).

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        if self.ident.is_anon() and self in self.parent._anonChildren:
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        yield from self._children
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Keep this.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py:341]: And this.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py:350]: And that

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py:357]: Keep this and put whatever you introduced in a method. People might want the debug information.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        symbol = parentSymbol._childrenByName.get(name)
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        return SymbolLookupResult([], parentSymbol, ident)
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py:504]: Shouldn't you add `otherChild` to `self._childrenByDocname` at some point? (I'm not really sure what this method does).

**Review by picnixz — DISMISSED** [review]

**Comment by jakobandersen** [comment]
It's great to see a PR addressing this. If you give me a week I should be able to review it as well. I'm a bit concerned with the simplified lookup logic as there are quite some corner cases.

**Comment by picnixz** [comment]
I'll wait for Jakob's review because I am not that well-versed in the C/C++ domain.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Good suggestions.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: I'll write a clearer comment for all three 'children' properties.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Instead of this, I will refactor into `_add_child` and `_remove_child` methods.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Refactored into `_add_child()`

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Good suggestion.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Good catch. Also changed `_childrenByDocname` to contain dicts so that child removal is done by name.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Refactored into a single return.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py:504]: Will refactor to use `_add_child()`

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Good suggestion. I will also simplify the method a bit more after this change.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [reply sphinx/domains/c/_symbol.py]: By the way, you should be able to remove the `identifier` all together since the module is private.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: The logic became messier when this was extracted into a method so I'd prefer to leave as is.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py:357]: I don't quite understand the ask here. `_find_named_symbols` covered several different lookup patterns and I removed it in favour of tailored code for each access pattern. I could add logging around the lookup, but I didn't touch logging at all in this PR because the logging calls are all broken if you enable it.

Comment (picnixz) [reply sphinx/domains/c/_symbol.py:357]: Oh, sorry I wasn't clear enough. Yes if possible, could you make similar logging calls or is your implementation not logging call friendly?

**Comment by donaldh** [comment]
> It's great to see a PR addressing this. If you give me a week I should be able to review it as well. I'm a bit concerned with the simplified lookup logic as there are quite some corner cases.

@jakobandersen As I understand it, the code came from the C++ domain where lookup complexity is higher due to namespaces and classes. I have tried my best to verify the lookup logic required for C. I verified the before and after generated output is the same for `make htmldocs` in the Linux kernel. If there are corner cases I have missed then we need to add them to the tests in sphinx.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Do you want to have a fixed sequence (i.e., is it possible that the child map is updated at the same time you are iterating over it from somewhere else?)

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]:
```suggestion
        if name in siblings:
            del siblings[name]
```

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py:189]: Maybe an empty line after the return (just for clarity)

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Since `childrenByDocname[docname]` does not change during the iteration (and you calling `del self._childrenByDocname[docname]`, I would suggest:

```python
children = self._childrenByDocname.pop(docname)
for child in children:
    # do the stuff
```

**Review by picnixz — COMMENTED** [review]
Small nitpicks.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Good suggestion, thanks.

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Are you asking why I have used `list(...)`? I got this error from `mypy`:

`Incompatible return value type (got "dict_values[str, Symbol]", expected "Sequence[Symbol]")`

We don't want to expose the dict behind the `_children` alias so I made it a list. Is there a better approach?

File: sphinx/domains/c/_symbol.py
Comment (donaldh) [reply sphinx/domains/c/_symbol.py:357]: The logging calls in `_symbol.py` are broken so I can't enable `.debug_lookup` to test any additions I make. It seems better to fix all that in a separate PR.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [reply sphinx/domains/c/_symbol.py]: The issue with mypy is because the type is a `Sequence` and `.values()` is an iterator view (more precisely a `ValuesView`), so you can change its typing.

However, when you use a `list`, you'll create a new object so if the are a lot of children, you can be *very* slow, e.g.

```python

import timeit
from random import random, randint

def y():
	return [random() for _ in range(randint(1000))]

x = {i: y() for i in range(1000)}

timeit.timeit('list(x.values())', globals={'x': x})
# 2.46

timeit.timeit('x.values()', globals={'x': x})
# 0.026
```

> We don't want to expose the dict behind the _children alias so I made it a list. Is there a better approach?

Are you talking about the possibility of writing `_children.mapping` ?

Comment (donaldh) [reply sphinx/domains/c/_symbol.py]: Okay, I'll change the type to `ValuesView[Symbol]`.

File: sphinx/domains/c/_symbol.py
Comment (picnixz) [inline sphinx/domains/c/_symbol.py]: Just wondering, but is it possible that the resolution could have different results if the anonymous children are not ordered in the order they were discovered?

Comment (jakobandersen) [reply sphinx/domains/c/_symbol.py]: It could, but I think in all cases then the original C code would not compile.
The following code illustrates it:
```rst

.. c:struct:: A

	:c:var:`a`

	.. c:union:: @U1

		:c:var:`a`

		.. c:var:: int a

			:c:var:`a`

	.. c:var:: int a

		:c:var:`a`

	.. c:union:: @U2

		:c:var:`a`

		.. c:var:: int a

			:c:var:`a`

:c:var:`A.a`
```
The very last xref will in master resolve to ``A.@U1.a`` because it is declared first, which is fine.
In this PR it resolves to ``A.a``, which in a sense is prettier.
The corresponding C code is something like
```c
struct A {
	union {
		int a;
	};
	int a;
	union {
		int a;
	};
};
```
which gives duplicate declaration errors.

**Review by jakobandersen — CHANGES_REQUESTED** [review]
Unfortunately my gut feeling was right, there is some handling of duplicate symbols which changes in this PR. 
Test case:
```rst
.. c:struct:: A

	.. c:var:: int i

	- :c:struct:`A`
	- :c:var:`A.i`
	- :c:var:`A.j`

.. c:struct:: A

	.. c:var:: int j

	- :c:struct:`A`
	- :c:var:`A.i`
	- :c:var:`A.j`

- :c:struct:`A`
- :c:var:`A.i`
- :c:var:`A.j`
```
When a duplicate symbol is detected, then the duplicate should not overwrite the original in the dictionaries. I think it is safe to throw the duplicate symbol away, as long as the old behaviour of setting the current scope to be the original symbol is kept.

**Comment by jakobandersen** [comment]
By the way, I fixed so the symbol debugging now works again, so a rebase onto master might be nice.

**Comment by donaldh** [comment]
> Unfortunately my gut feeling was right, there is some handling of duplicate symbols which changes in this PR. Test case:
> [...] 
> When a duplicate symbol is detected, then the duplicate should not overwrite the original in the dictionaries. I think it is safe to throw the duplicate symbol away, as long as the old behaviour of setting the current scope to be the original symbol is kept.

I think this is caused by `candSymbol` in `_add_symbols()`. A new symbol with a parent gets automatically added to the parent's `_childrenBy*` so this happens for `candSymbol` before the duplicate checking gets run. I did look at removing the `_add_child()` step from `__init__` and doing it after duplicate checking, but this breaks other code paths. Instead, I think it is sufficient to check for a name collision in `_add_symbols()` and never overwrite an existing entry:

```
diff --git a/sphinx/domains/c/_symbol.py b/sphinx/domains/c/_symbol.py
index 0c8905f1d59f..5194427b5466 100644
--- a/sphinx/domains/c/_symbol.py
+++ b/sphinx/domains/c/_symbol.py
@@ -123,6 +123,9 @@ class Symbol:
 
     def _add_child(self, child: Symbol) -> None:
         name = child.ident.name
+        if name in self._childrenByName:
+            # Duplicate so don't add - will be reported in _add_symbols()
+            return
         self._childrenByName[name] = child
         if child.docname not in self._childrenByDocname:
             self._childrenByDocname[child.docname] = {}
```

**Comment by donaldh** [comment]
While investigating another issue, I have found something that the C domain does not account for. I was looking for the root cause of these false duplicate warnings we get from Sphinx 3.0.0 thru 7.6.2:
```
/home/donaldh/net-next/Documentation/driver-api/usb/usb:164: /home/donaldh/net-next/drivers/usb/core/message.c:968: WARNING: Duplicate C declaration, also defined at driver-api/usb/gadget:802.
Declaration is '.. c:function:: int usb_string (struct usb_device *dev, int index, char *buf, size_t size)'.
/home/donaldh/net-next/Documentation/driver-api/usb/usb.rst:968: WARNING: Duplicate C declaration, also defined at driver-api/usb/gadget:802.
Declaration is '.. c:struct:: usb_string'.
/home/donaldh/net-next/Documentation/driver-api/miscellaneous:47: /home/donaldh/net-next/drivers/pwm/core.c:534: WARNING: Duplicate C declaration, also defined at driver-api/miscellaneous:224.
Declaration is '.. c:function:: int pwm_capture (struct pwm_device *pwm, struct pwm_capture *result, unsigned long timeout)'.
/home/donaldh/net-next/Documentation/gpu/drm-kms:360: /home/donaldh/net-next/drivers/gpu/drm/drm_fourcc.c:344: WARNING: Duplicate C declaration, also defined at gpu/drm-kms:39.
Declaration is '.. c:function:: const struct drm_format_info * drm_format_info (u32 format)'.
/home/donaldh/net-next/Documentation/gpu/drm-kms:461: /home/donaldh/net-next/drivers/gpu/drm/drm_modeset_lock.c:392: WARNING: Duplicate C declaration, also defined at gpu/drm-kms:49.
Declaration is '.. c:function:: int drm_modeset_lock (struct drm_modeset_lock *lock, struct drm_modeset_acquire_ctx *ctx)'.
/home/donaldh/net-next/Documentation/gpu/drm-uapi:434: /home/donaldh/net-next/drivers/gpu/drm/drm_ioctl.c:874: WARNING: Duplicate C declaration, also defined at gpu/drm-uapi:70.
Declaration is '.. c:function:: bool drm_ioctl_flags (unsigned int nr, unsigned int *flags)'.
/home/donaldh/net-next/Documentation/core-api/workqueue:776: /home/donaldh/net-next/include/linux/workqueue.h:493: WARNING: Inline literal start-string without end-string.
```
The root cause is e.g. `struct usb_string` and `int usb_string()` both being defined, which the C domain doesn't account for. Sphinx is treating this as a duplicate when it shouldn't. They should be kept distinct, within separate search spaces. One way to resolve this would be to introduce two separate search spaces:
1. Tag names – structs, unions, enums.
2. Identifier names – functions, vars, typedefs.
Question: Should I tackle that in this PR or in a separate PR?

**Comment by jakobandersen** [comment]
> 1. Tag names – structs, unions, enums.
> 2. Identifier names – functions, vars, typedefs.
> 
> Question: Should I tackle that in this PR or in a separate PR? The reason I ask is because it will affect the implementation used in this PR.

There is an old issue for this (https://github.com/sphinx-doc/sphinx/issues/7819), and I have a branch that should fix it (https://github.com/jakobandersen/sphinx/tree/c_ids_tags). As soon as I get time I will rebase it and get it into shape. It has been a while, but if I remember correctly it was very difficult to not break intersphinx support without https://github.com/sphinx-doc/sphinx/pull/8929. I'm currently updating it, so there can be progress.

**Comment by jakobandersen** [comment]
> I did look at removing the `_add_child()` step from `__init__` and doing it after duplicate checking, but this breaks other code paths. Instead, I think it is sufficient to check for a name collision in `_add_child()` and never overwrite an existing entry

I tried the same quick fix, and indeed it broke some invariants elsewhere. I think your solution for now it's a good way to fix it.

**Comment by donaldh** [comment]
Hi @jakobandersen, @picnixz is there anything left that I need to resolve? I can't spot any open issues in the thread above. It would be great to get this merged so please let me know if there is more needed from me.

**Comment by picnixz** [comment]
I personally am not well-versed enough in the C/C++ domain so I'll leave it to Jakob (would have been another question if it were the Python domain I'm most familiar with :smile:)

**Comment by jakobandersen** [comment]
I hope to get to it in the coming weekend.

**Review by jakobandersen — APPROVED** [review]

**Comment by jakobandersen** [comment]
Looks good to me. Apparently I can't merge it without either squashing or rebasing, is that intentional?

**Comment by picnixz** [comment]
> Looks good to me. Apparently I can't merge it without either squashing or rebasing, is that intentional?

I think our policy is to only to squash commits + write description if needed instead of rebasing + possible multiple commits.

**Comment by jakobandersen** [comment]
> I think our policy is to only to squash commits + write description if needed instead of rebasing + possible multiple commits (although it's possible to rebase but I don't think we've ever merged PRs by rebasing them). But we don't want to merge the user's commits since generally there are not.. like.. explicit enough (like mine !)

Must be a "new" policy (in the last couple of years :-)). I see the point, but on the surface it doesn't seem like a good idea in general to destroy the history of changes made.
Anyway, from my side please go ahead with merging this (in some way).

**Comment by picnixz** [comment]
> Must be a "new" policy (in the last couple of years :-)).

For me it's the policy since I only arrived last year haha! But I can understand that you would like not to destroy the commits. But I think it should be fine to rebase? (and keep the commits?) 

@AA-Turner Any reason why we don't allow just using a commit message? (or can we use rebase?)

**Comment by AA-Turner** [comment]
@donaldh please could you add an entry to CHANGES for 7.4?

**Review by AA-Turner — APPROVED** [review]
All good, save the CHANGES entry.

A

**Comment by AA-Turner** [comment]
Thank you @donaldh!

A

**Comment by AA-Turner** [comment]
@donaldh -- would you consider making a similar PR for the C++ domain? I have had a stab at replicating your work in https://github.com/AA-Turner/sphinx/tree/cpp/lookup (https://github.com/AA-Turner/sphinx/commit/109ec9cdfb9cb8e083be4c732171064a964254ec), but my knowledge of C++ is limited.

Importantly, the C++ domain uses ``Identifier | Operator`` for the types, and ``Operator`` doesn't have a ``name`` attribute.

A

**Comment by donaldh** [comment]
@AA-Turner I will try to find some time this week to take a look.

**Pull request merged**
